src/object/obj_class.c

```diff
@@ -114,7 +114,7 @@ static struct daos_obj_class daos_obj_classes[] = {
 		{
 			.ca_schema		= DAOS_OS_STRIPED,
 			.ca_resil		= DAOS_RES_REPL,
-			.ca_grp_nr		= 2,
+			.ca_grp_nr		= DAOS_OBJ_GRP_MAX,
```

Comment:
Not sure if this is what we want for replica_3 object, since it will create a big layout for all replica_3 object, probably add a replica_3_max object? Though I agree the object class definition is a bit messy for now.

In the mean time probably we also need to output per VOS pool size, so the user can "understand" why it can get ENOSPC.

Reply:
there is already a R3S oclass for small.
So i think for R3, this is appropriate, at least from a user perspective.
As we are running those tests, it didn't occur to me that this is an issue until i did a dmg layout and saw the problem, so no it 's not expected from user perspective that ENOSPC will be returned with R3 given there is more than enough space on all targets.

Reply:
What I mean is 
1. either you change the object class name to repl_3_max_rw ? 
2 or keep the current object class but maybe change the name to repl_3_grp_2_rw, then add another object class as repl_3_max_rw?

Just to be clear, right now we generate object layout based on VOS target, instead of based  on server as before.  i.e. it will only write to a single vos target (6.5GB) for each server (assume 1 server == 1 fault domain), instead of all VOS targets on this server(50GB) as before.  I think that is why the ENOSPC is met here.

Reply:
Oh i am perfectly aware why we met ENOSPC.
Im just curious why you need this repl_3_grp_2_rw object class? is there a specific use case for it? from user perspective i don't see any other than for simple testing if it's need in the tests..

Reply:
First, max is only for testing purpose, and it simply use all of targets of the pool to create the layout. for example 3 servers,  each server has 10 targets, if creating a 3 repl obj with max grp_nr, it simply generate 3-way replica on all 3x10 targets, which does not consider fault domain at all.  i.e. this is almost "useless" in the real world IMHO.  This is why I do not want to change grp_nr of repl_3_rw to be "max" always. So just add repl_3_max_rw for testing purpose?

As for why set grp to be 2 for repl_3_rw, TBH, I do not care if it is 2 , 3, 4 or other number, but I do care the object name does not reflect the group number, that is why I suggest to add grp_2 inside this name.

Reply:
i guess i don't understand why you need grp_2/3/4 etc. if you don't care if it's 2, 3, 4 or whatever. if this is dynamic then it should be a user defined class, and not exported as one of the classes here.
So i think by default this should be max, even for testing purposes and not real world scenario as you claim.

Reply:
we do not support user defined "grp_nr" for object class, that is why we use static definition for now.  So we use object id to define the width of the object layout.

In the real world, when defining the object layout, you have to consider the fault domain,  I do not think normally the user will use all targets to generate the object layout, maybe in  some extrem cases, definitely not "default" behavior, IMHO.

Reply:
well i don't want to change the API by renaming this object class and we need R3 to use all targets today because we do not have enough space for our testing to use a subset of the targets. TBH i don't understand why you are arguing against this patch. If a specific group nr is needed, the object class should be renamed using that nr in it's name. all i did was change the R3 object class to use max grp nr because that's what we need. i think that is the appropriate setting for just R3.

we can definitely add the r3_grp_x oclass, but that can be done in a different patch and for the actual grp_nr that is needed. im not arguing against that.

Reply:
All I am saying here is that do not set max as the default grp_nr for R3, since this does not make sense, and also cause confusion for others, IMHO. If you want to use max grp_nr create another one with specific name. 

What API do you mean here?  I am sorry if I miss sth.

Reply:
Well there is no default.. you would either use r3 with all targets or a specific r3 with a certain number of grp. I don’t see a default setting in object class. There is no confusion.. 
For api, I mean the o class name. It’s used on other middleware and we don’t want to trigger a change in name that breaks all those.

Reply:
I just don’t want to have to change mpiio DFS and HDF5. But if everyone thinks it’s a required change for all object class to include the grp nr and max for all, then it’s fine, we can make that change

Reply:
Sorry, for me this is user level api change,  I do not want cause any confusion(either to user or developer) here. Maybe Liang or Xuezhao could post their opinion here. Sorry to be picky here.

Reply:
Np, I understand, but please be aware all oclass name would have to change in this case to include the grp nr if you want to go with that including large. I want to avoid doing that

Reply:
Hi, I can understand Di's concern. 
Using DAOS_OBJ_GRP_MAX as default grp_nr for DAOS_OC_R3_RW will cause a problem that when there is node failure, it possibly cannot find spare target (assume it uses all targets). you can refer the code about ring_obj_remap_shards().
So can say that the MAX GROUP commonly only used for performance testing purpose. If you want to use it seems reasonable to me that should add a special obj class named with _max_.

On the other hand, I can see Mohamod's requirement that one MPI shared file, want to use more (or even all) available targets. I am not very sure, maybe can provide a few options/hints to user that possibly without fault-tolerance for tgt/node failure, or use another obj class that with most possibe largest number of tgt but still keep a few tgts as possible spare tgts, or not sure if it is possible to map one MPI shared file to multiple DAOS objects so can use more/all tgts.

Reply:
"one MPI shared file, want to use more (or even all) available targets", or another option that change a few placement code (ring_obj_remap_shards related code), to allow it select some already used tgt as spare tgt for the _max obj class. May see others comments/suggestions as well. thx

Reply:
we need to resolve it for v0.6.
For DAOS_OBJ_GRP_MAX, I think there are two issues:
1) as the obj layout already covered all tgts, it cannot find spare targets for rebuild when some one fails.
2) it is hard to handle the case when new server/tgts adds to pool, the actual grp_nr will increase in the new pool map ver, so the dkeys location will be changed. So fetch data in new map ver while the data updated on old map ver will meet problem (cannot wait until data rebalanced).

@mchaarawi , Is it possible to totally remove DAOS_OBJ_GRP_MAX definition and usage? can we satisfy upper layer's requirement by other approach? ( for example map the mpi files to multiple objects, but i am not sure with it).